### ISSUE TITLE
masking prediction help

### ISSUE THREAD

**Issue by danpf** [agemagician/ProtTrans#31]
Hello,
Thank you for providing the prottrans models / examples here they are very helpful!

I'm playing around with predicting sequences from masks and I'm getting strange results. I wonder if you could lead me in the right direction.
Using the following code:

<details>
  <summary>example code</summary>

```
from transformers import BertForMaskedLM, BertTokenizer, pipeline
import re
import os

import torch
import requests
from tqdm.auto import tqdm


def x_to_mask(sequences):
    # sequences = [" ".join(["[CLS]"] + list(x) + ["[SEP]"]) for x in sequences]
    sequences = [" ".join(list(x)) for x in sequences]
    sequences = [re.sub(r"[X]", "[MASK]", x) for x in sequences]
    sequences = [re.sub(r"[UZOB]", "X", x) for x in sequences]
    return sequences


def mask_range(device, model, tokenizer):
    correct_seq = "AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR"
    input_sequences = [
        "AQVINTFDGVADYXXXXXXXXXXXXXXXXXQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLAXXXXXXXXXXXIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSXXXXXXXXXXXXXGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDXXXXXXXXX",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQXLGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQXXGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQXXXWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQXLGWVASKGNLADVAPGKSIGGXIFSNREGKLPGKSGRTWREADINYTSXFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKXEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREXDINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR",
        "AQVINTFDGVADYXQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSXDWLIYKTTDHYQTFTKIR",
        ]
    sequences = x_to_mask(input_sequences)
    tokseq = tokenizer(sequences, return_tensors="pt").to(device)
    labels = tokenizer(x_to_mask([correct_seq])*len(sequences), return_tensors="pt").to(device)
    results = model(**tokseq, labels=labels["input_ids"])
    print("loss", results.loss)
    preds = ["".join(tokenizer.convert_ids_to_tokens(x[1:-1])) for x in torch.argmax(results.logits, dim=2)]
    for i in range(len(sequences)):
        print(f"attempt {i} input/pred/correct")
        print(input_sequences[i])
        print(preds[i])
        print(correct_seq)
        checker = []
        for j in range(len(correct_seq)):
            if correct_seq[j] != preds[i][j]:
                checker.append("*")
            else:
                checker.append("-")
        checker_str = "".join(checker)
        print(checker_str)
        print(f"wrong = {checker_str.count('*')} / {input_sequences[i].count('X')}")
    import pdb
    pdb.set_trace()


def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    tokenizer = BertTokenizer.from_pretrained('Rostlab/prot_bert_bfd', do_lower_case=False )
    model = BertForMaskedLM.from_pretrained("Rostlab/prot_bert_bfd").to(device)
    model.eval()
    with torch.no_grad():
        mask_range(device, model, tokenizer)

if __name__ == "__main__":
    main()

```

</details>

I am ending up not getting very much correct -- and even some unmasked residues wrong  Is there anything obvious that I'm doing wrong?

sample output
<details>
  <summary>sample output</summary>

```
loss tensor(1.0439, device='cuda:0')
attempt 0 input/pred/correct
AQVINTFDGVADYXXXXXXXXXXXXXXXXXQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYAAAGGGGGGGAAGGGGGQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------*****************---------------------------------------------------------------**------**-------
wrong = 21 / 17
attempt 1 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLAXXXXXXXXXXXIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLALLLDDDGGGGLIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDHGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------------------********--*---------------------------------------**-------*-------
wrong = 12 / 11
attempt 2 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSXXXXXXXXXXXXXGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSLLGGGGGGGGIAIGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------------------------------------------*************-------------**------**-------
wrong = 17 / 13
attempt 3 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDXXXXXXXXX
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDGKGKLLLGX
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-----------------------------------------------------------------------------------------------------*********
wrong = 9 / 9
attempt 4 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQXLGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQKLGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------*-------------------------------------------------------------**------**-------
wrong = 5 / 1
attempt 5 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQXXGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQKLGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------*-------------------------------------------------------------**------**-------
wrong = 5 / 2
attempt 6 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQXXXWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQLLLWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDHGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------*-*-----------------------------------------------------------**-------*-------
wrong = 5 / 3
attempt 7 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQXLGWVASKGNLADVAPGKSIGGXIFSNREGKLPGKSGRTWREADINYTSXFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQKLGWVASKGNLADVAPGKSIGGLIFSNREGKLPGKSGRTWREADINYTSIFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------*---------------------*--------------------------*------------**------**-------
wrong = 7 / 3
attempt 8 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKXEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKEEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
---------------------------*-----------------------------------------------------------------**------**-------
wrong = 5 / 1
attempt 9 input/pred/correct
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREXDINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREGDINYTSGFRNSDRILYSSDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------------------------------------------------*-------------------**------**-------
wrong = 5 / 1
attempt 10 input/pred/correct
AQVINTFDGVADYXQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSXDWLIYKTTDHYQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSTDGGIYKTTDGGQTFTKIR
AQVINTFDGVADYLQTYHKLPDNYITKSEAQALGWVASKGNLADVAPGKSIGGDIFSNREGKLPGKSGRTWREADINYTSGFRNSDRILYSSDWLIYKTTDHYQTFTKIR
-------------------------------------------------------------------------------------------*-**------**-------
wrong = 5 / 2

```

</details>

do these results make sense to you at all? I'm guessing it's just my implementation but I'm not sure where I have gone wrong

*EDIT*
This is almost certainly due to me using 1) multiple masks, or/and 2) not using the fill-mask pipeline (which is only able to use one mask at a time)... Perhaps the question I should be asking is -- how can I properly fill-multiple-masks?

**Comment by agemagician** [comment]
We have not tested multiple mask filling, we only tested single mask filling, which did work as expected by using fill-mask pipeline.
It is better to ask this question in hugging face forum:
https://discuss.huggingface.co

I have found this on their forum:
https://discuss.huggingface.co/t/multiple-mask-tokens/174
It should give you some direction.
